File: instantlib/src/main/java/com/baselib/instant/manager/ObserverManager.java
package com.baselib.instant.manager;

import android.content.Context;

import com.baselib.instant.observer.observer.AppChangeObserver;
import com.baselib.instant.observer.IObserver;
import com.baselib.instant.util.LogUtils;

import java.util.HashMap;
import java.util.Map;

/**
 * 观察者管理对象
 *
 * @author wsb
 */
public class ObserverManager implements IManager {

    private static final String APP_CHANGE_OBSERVER_NAME = "AppChangeObserver";
    private Map<String, IObserver> mObserverMap;

    public ObserverManager() {
        mObserverMap = new HashMap<>();
    }

    @Override
    public void detach() {
        for (String name : mObserverMap.keySet()) {
            IObserver observer = mObserverMap.get(name);
            if (observer != null) {
                observer.onDestroy();
            }
        }
        mObserverMap = null;
    }

    /**
     * 该方法内对观察者管理对象进行初始化
     * <p>
     * 后面get方法获取到的观察者对象不为空需要在此put进容器钟
     *
     * @param context 上下文
     */
    public void attach(Context context) {
        mObserverMap.put(APP_CHANGE_OBSERVER_NAME, new AppChangeObserver(context));
    }

    public AppChangeObserver getAppChangeObserver() {
        return getObserver(APP_CHANGE_OBSERVER_NAME, AppChangeObserver.class);
    }

    private <T extends IObserver> T getObserver(String name, Class<T> clz) {
        IObserver observer = mObserverMap.get(name);
        if (observer == null) {
            LogUtils.d("该观察者暂未被支持");
        }
        return observer == null ? null : (T) observer;
    }
}

File: instantlib/src/main/java/com/baselib/instant/observer/ObserverManager.kt
package com.baselib.instant.observer

import android.content.Context

import com.baselib.instant.manager.IManager
import com.baselib.instant.observer.observer.AppChangeObserver
import com.baselib.instant.observer.observer.BaseObserver
import com.baselib.instant.observer.observer.NetStateObserver
import com.baselib.instant.util.LogUtils

import java.util.HashMap

/**
 * 观察者管理对象
 *
 * @author wsb
 */
class ObserverManager : IManager {
    private var mObserverMap: MutableMap<String, IObserver<*>> = HashMap()

    override fun onManagerDetach() {
        for (name in mObserverMap.keys) {
            mObserverMap[name]?.onObserverDetach()
        }
        mObserverMap.clear()
    }

    fun <T : IObserver<*>> getObserver(context: Context, name: String): T? {
        return if (null == mObserverMap[name]) {
            createObserver(context, name).takeIf { it != null }?.also {
                mObserverMap[name] = it
            } as T?
        } else {
            mObserverMap[name] as T
        }
    }

    private fun createObserver(context: Context, name: String): IObserver<*>? {
        return when (name) {
            APP_CHANGE_OBSERVER_NAME -> AppChangeObserver(context)
            NET_STATE_OBSERVER_NAME -> NetStateObserver(context)
            else -> null
        }
    }

    companion object {
        const val APP_CHANGE_OBSERVER_NAME = "AppChangeObserver"
        const val NET_STATE_OBSERVER_NAME = "NetStateObserver"
    }
}
